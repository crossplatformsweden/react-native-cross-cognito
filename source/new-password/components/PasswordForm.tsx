import React from 'react';
import { View } from 'react-native';
import {
  CrossButton,
  Colors,
  ICrossButtonProps,
} from 'react-native-cross-components';
import styles from '../../styles';
import { CrossEditor } from 'react-native-cross-components';
import {
  IEmailProps,
  IPasswordProps,
  ICodeProps,
  ITestIDProps,
} from '../../types';

/**
 * Props for the {@link PasswordForm} component.
 *
 * Required properties: {@link onSavePress}.
 *
 * Allows you to customize {@link ICodeProps}, {@link IEmailProps} and {@link IPasswordProps}.
 */
export interface IPasswordFormProps
  extends IEmailProps,
    IPasswordProps,
    ICodeProps,
    ITestIDProps {
  /**
   * Occurs when the user press "save" button. Required.
   */
  onSavePress: () => void;
  /**
   * Optional props for the register button. Typically used to change the `title` prop.
   *
   * Read more:
   * https://crossplatformsweden.github.io/react-native-components/interfaces/_components_buttons_crossbutton_.icrossbuttonprops.html
   *
   * @example
   *  <PasswordForm confirmButtonProps={{title: 'Enlist'}}>
   *     <Text>Custom layouts here</Text>
   *  </PasswordForm>
   */
  saveButtonProps?: ICrossButtonProps | undefined;
}

/**
 * Allows the user to ender credentials and passes them back through prop events.
 *
 * Props are {@link IPasswordFormProps}
 *
 * @example
 *  <PasswordForm
 *      onSavePress={(code) => console.log('Confirmed', code)}
 *      saveButtonProps={{title: 'Enlist'}}
 *      codeInputProps={{label: 'WouldntYouLikeToKnow'}}>
 *     <Text>Custom layouts here</Text>
 *  </PasswordForm>
 */
export class PasswordForm extends React.Component<IPasswordFormProps> {
  render() {
    return (
      <View style={[styles.container]} testID={this.props.testID}>
        {this.props.children}{' '}
        <CrossEditor
          testID='email'
          label='E-mail'
          placeholder='E-mail'
          autoCapitalize='none'
          clearButtonMode='always'
          autoCorrect={false}
          value={this.props.initialEmail || ''}
          onChangeText={(text: string) => this.props.onEmailChanged(text)}
          {...this.props.emailInputProps}
        />
        <CrossEditor
          testID='code'
          label='Confirmation code'
          placeholder='Enter the code that was sent to you'
          autoCapitalize='none'
          clearButtonMode='always'
          autoCorrect={false}
          value={this.props.code || ''}
          onChangeText={this.props.onCodeChanged}
          {...this.props.codeInputProps}
        />
        <CrossEditor
          testID='password'
          label='New Password'
          clearButtonMode='always'
          secureTextEntry
          value={this.props.initialPassword || ''}
          onChangeText={(text: string) => this.props.onPasswordChanged(text)}
          {...this.props.passwordInputProps}
        />
        <CrossButton
          style={styles.marginTop10}
          buttonStyle={styles.buttonStyle}
          onPress={async () => await this.props.onSavePress()}
          mode='contained'
          title='Save'
          backgroundColor={Colors.NextButton}
          iconName='sign-in'
          {...this.props.saveButtonProps}
        />
      </View>
    );
  }
}

export default PasswordForm;
