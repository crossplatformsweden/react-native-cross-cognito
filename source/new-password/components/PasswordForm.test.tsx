// https://itnext.io/testing-react-16-3-components-with-react-test-renderer-without-enzyme-d9c65d689e88
import React from 'react';
import TestRenderer from 'react-test-renderer';
import PasswordForm from './PasswordForm';

jest.unmock('react-native');
jest.unmock('./PasswordForm');

// Default export requires this type of mocking
// jest.mock('react-native-stager', () => ({
//   __esModule: true,
//   Stage: 'View',
//   StageButtons: 'View',
//   StageProgress: 'View',
//   default: 'View',
// }));

describe('components', () => {
  describe('<PasswordForm />', () => {
    it('Component should render', () => {
      const wrapper = TestRenderer.create(
        <PasswordForm
          onPasswordChanged={console.log}
          code='testcode'
          onSavePress={console.log}
          onEmailChanged={console.log}
        />
      );
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('`onEmailChanged` should return string', () => {
      let result: string | undefined = undefined;
      const onCalled = (r: string | undefined) => (result = r);
      const wrapper = TestRenderer.create(
        <PasswordForm
          onPasswordChanged={console.log}
          code='testcode'
          onSavePress={console.log}
          onEmailChanged={onCalled}
        />
      );
      const child = wrapper.root.findByProps({ label: 'E-mail' });
      child.props.onChangeText('testcode');

      expect(result).toBe('testcode');
    });

    it('`onPasswordChanged` should return string', () => {
      let result: string | undefined = undefined;
      const onCalled = (r: string | undefined) => (result = r);
      const wrapper = TestRenderer.create(
        <PasswordForm
          onEmailChanged={console.log}
          code='testcode'
          onSavePress={console.log}
          onPasswordChanged={onCalled}
        />
      );
      const child = wrapper.root.findByProps({ label: 'New Password' });
      child.props.onChangeText('testcode');

      expect(result).toBe('testcode');
    });

    it('`onCodeChanged` should return string', () => {
      let result: string | undefined = undefined;
      const onCalled = (r: string | undefined) => (result = r);
      const wrapper = TestRenderer.create(
        <PasswordForm
          onEmailChanged={console.log}
          onPasswordChanged={console.log}
          code='testcode'
          onSavePress={console.log}
          onCodeChanged={onCalled}
        />
      );
      const child = wrapper.root.findByProps({
        label: 'Confirmation code',
      });
      child.props.onChangeText('testcode');

      expect(result).toBe('testcode');
    });

    it('<CrossButton /> onPress should be called', () => {
      let called = false;
      const onCalled = () => (called = !called);

      const wrapper = TestRenderer.create(
        <PasswordForm
          onPasswordChanged={console.log}
          code='testcode'
          onSavePress={onCalled}
          onEmailChanged={console.log}
        />
      );
      const child = wrapper.root.findByProps({ title: 'Save' });
      child.props.onPress();
      expect(called).toBeTruthy();
    });
  });
});
