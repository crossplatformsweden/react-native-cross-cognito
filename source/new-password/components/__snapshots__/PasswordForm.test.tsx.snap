// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <PasswordForm /> Component should render 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
   
  <TextInput
    autoCapitalize="none"
    autoCorrect={false}
    clearButtonMode="always"
    label="E-mail"
    onChangeText={[Function]}
    placeholder="E-mail"
    value=""
  />
  <TextInput
    autoCapitalize="none"
    autoCorrect={false}
    clearButtonMode="always"
    label="Confirmation code"
    placeholder="Enter the code that was sent to you"
    value="testcode"
  />
  <TextInput
    clearButtonMode="always"
    label="New Password"
    onChangeText={[Function]}
    secureTextEntry={true}
    value=""
  />
  <Button
    backgroundColor="#349b34"
    buttonStyle={
      Object {
        "margin": 0,
        "minHeight": 45,
      }
    }
    iconName="sign-in"
    mode="contained"
    onPress={[Function]}
    style={
      Object {
        "marginTop": 10,
      }
    }
    title="Save"
  />
</View>
`;
